**CLEANING Rates_of_Cases_or_Deaths.csv** (First dataset)

In [3]:
import pandas as pd
import shutil
import re
import numpy as np

In [2]:
raw = pd.read_csv("Data/raw/Rates_of_Cases_or_Deaths.csv")

In [3]:
print(raw)

     outcome     month  MMWR week     Age group Vaccine product  \
0       case  APR 2021     202114         12-17       all_types   
1       case  APR 2021     202114         18-29       all_types   
2       case  APR 2021     202114         30-49       all_types   
3       case  APR 2021     202114         50-64       all_types   
4       case  APR 2021     202114         65-79       all_types   
...      ...       ...        ...           ...             ...   
1586   death  AUG 2022     202235           80+       all_types   
1587   death  AUG 2022     202235  all_ages_adj       all_types   
1588   death  AUG 2022     202235  all_ages_adj         Janssen   
1589   death  AUG 2022     202235  all_ages_adj         Moderna   
1590   death  AUG 2022     202235  all_ages_adj          Pfizer   

      Vaccinated with outcome  Fully vaccinated population  \
0                           8                 3.688700e+04   
1                         674                 2.543093e+06   
2        

HANDLING MISSING AND INVALID DATA

In [7]:
# number of missing values(data)
raw.isnull()

outcome  month  MMWR week  Age group  Vaccine product  \
0       False  False      False      False            False   
1       False  False      False      False            False   
2       False  False      False      False            False   
3       False  False      False      False            False   
4       False  False      False      False            False   
...       ...    ...        ...        ...              ...   
1586    False  False      False      False            False   
1587    False  False      False      False            False   
1588    False  False      False      False            False   
1589    False  False      False      False            False   
1590    False  False      False      False            False   

      Vaccinated with outcome  Fully vaccinated population  \
0                       False                        False   
1                       False                        False   
2                       False                        False   
3                       False                        False   
4                       False                        False   
...                       ...                          ...   
1586                    False                        False   
1587                    False                        False   
1588                    False                        False   
1589                    False                        False   
1590                    False                        False   

      Unvaccinated with outcome  Unvaccinated population  Crude vax IR  \
0                         False                    False         False   
1                         False                    False         False   
2                         False                    False         False   
3                         False                    False         False   
4                         False                    False         False   
...                         ...                      ...           ...   
1586                      False                    False         False   
1587                      False                    False         False   
1588                      False                    False         False   
1589                      False                    False         False   
1590                      False                    False         False   

      Crude unvax IR  Crude IRR  Age adjusted vax IR  Age adjusted unvax IR  \
0              False      False                 True                   True   
1              False      False                 True                   True   
2              False      False                 True                   True   
3              False      False                 True                   True   
4              False      False                 True                   True   
...              ...        ...                  ...                    ...   
1586           False      False                 True                   True   
1587           False      False                False                  False   
1588           False      False                False                  False   
1589           False      False                False                  False   
1590           False      False                False                  False   

      Age adjusted IRR  Continuity correction  
0                 True                  False  
1                 True                  False  
2                 True                  False  
3                 True                  False  
4                 True                  False  
...                ...                    ...  
1586              True                  False  
1587             False                  False  
1588             False                  False  
1589             False                  False  
1590             False                  False  

[1591 rows x 16 columns]

In [8]:
raw.isnull().sum()

outcome                          0
month                            0
MMWR week                        0
Age group                        0
Vaccine product                  0
Vaccinated with outcome          0
Fully vaccinated population      0
Unvaccinated with outcome        0
Unvaccinated population          0
Crude vax IR                     0
Crude unvax IR                   0
Crude IRR                      114
Age adjusted vax IR            987
Age adjusted unvax IR          987
Age adjusted IRR               987
Continuity correction            0
dtype: int64

In [9]:
# % of missing values in each column: number of missing values/total values * 100
raw.isnull().sum()/len(raw)*100

outcome                         0.000000
month                           0.000000
MMWR week                       0.000000
Age group                       0.000000
Vaccine product                 0.000000
Vaccinated with outcome         0.000000
Fully vaccinated population     0.000000
Unvaccinated with outcome       0.000000
Unvaccinated population         0.000000
Crude vax IR                    0.000000
Crude unvax IR                  0.000000
Crude IRR                       7.165305
Age adjusted vax IR            62.036455
Age adjusted unvax IR          62.036455
Age adjusted IRR               62.036455
Continuity correction           0.000000
dtype: float64

In [40]:
## Age adjusted vax IR, Age adjusted unvax IR, and Age adjusted IRR have a high % of missing values. 
## We will delete them. As they are important for the analysis, we will replace them by age group metrics.
clean = raw.drop(["Age adjusted vax IR", "Age adjusted unvax IR", "Age adjusted IRR"], axis=1)
print(clean)

     outcome     month  MMWR week     Age group Vaccine product  \
0       case  APR 2021     202114         12-17       all_types   
1       case  APR 2021     202114         18-29       all_types   
2       case  APR 2021     202114         30-49       all_types   
3       case  APR 2021     202114         50-64       all_types   
4       case  APR 2021     202114         65-79       all_types   
...      ...       ...        ...           ...             ...   
1586   death  AUG 2022     202235           80+       all_types   
1587   death  AUG 2022     202235  all_ages_adj       all_types   
1588   death  AUG 2022     202235  all_ages_adj         Janssen   
1589   death  AUG 2022     202235  all_ages_adj         Moderna   
1590   death  AUG 2022     202235  all_ages_adj          Pfizer   

      Vaccinated with outcome  Fully vaccinated population  \
0                           8                 3.688700e+04   
1                         674                 2.543093e+06   
2        

In [41]:
# Save clean to csv file
clean.to_csv("clean_data_1.csv", index=False)

In [47]:
source = "clean_data_1.csv"
destination = "Data/clean/"
shutil.move(source, destination)

'Data/clean/clean_data_1.csv'

In [5]:
clean = pd.read_csv("Data/clean/clean_data_1.csv")

In [22]:
## Filling "Crude IRR" as % of missing values is 7.1653:
## Calculate mean of this column and fill the missing values with the mean.
mean_value = clean["Crude IRR"].mean()
print(mean_value)

7.014692613974949


In [23]:
# standardize missing values and fillin
clean['Crude IRR'] = clean['Crude IRR'].replace(['', ' ', 'NA', 'N/A', 'null'], np.nan)
print(clean['Crude IRR'].isnull().sum()) 
mean_value = clean['Crude IRR'].astype(float).mean()  # Ensure the column is numeric
clean['Crude IRR'] = clean['Crude IRR'].fillna(mean_value)
clean.to_csv('clean_data_1_updated.csv', index=False)

# Verify again in pandas
print(clean[clean['Crude IRR'].isnull()])  # Should return an empty DataFrame

114
Empty DataFrame
Columns: [outcome, month, MMWR week, Age group, Vaccine product, Vaccinated with outcome, Fully vaccinated population, Unvaccinated with outcome, Unvaccinated population, Crude vax IR, Crude unvax IR, Crude IRR, Continuity correction]
Index: []


In [24]:
source = "clean_data_1_updated.csv"
destination = "Data/clean/"
shutil.move(source, destination)

'Data/clean/clean_data_1_updated.csv'

In [51]:
clean.isnull().sum()

outcome                        0
month                          0
MMWR week                      0
Age group                      0
Vaccine product                0
Vaccinated with outcome        0
Fully vaccinated population    0
Unvaccinated with outcome      0
Unvaccinated population        0
Crude vax IR                   0
Crude unvax IR                 0
Crude IRR                      0
Continuity correction          0
dtype: int64

In [57]:
# Check for invalid entries in the outcome column
valid_values = ['case', 'death']
invalid_entries = clean[~clean['outcome'].isin(valid_values)]

# Display rows with invalid entries (if any)
print("Invalid entries:")
print(invalid_entries)

# Remove rows with invalid entries
clean[clean['outcome'].isin(valid_values)].copy()

# Verify the cleaning process
print("Unique values after cleaning:", clean['outcome'].unique())

Invalid entries:
Empty DataFrame
Columns: [outcome, month, MMWR week, Age group, Vaccine product, Vaccinated with outcome, Fully vaccinated population, Unvaccinated with outcome, Unvaccinated population, Crude vax IR, Crude unvax IR, Crude IRR, Continuity correction]
Index: []
Unique values after cleaning: ['case' 'death']


In [58]:
# Check for invalid entries in the month column.
# Define a valid pattern for month-year
def is_valid_month(value):
    return bool(re.match(r"^(Jan|Feb|Mar|Apr|May|June|July|Aug|Sept|Oct|Nov|Dec)(2021|2022)$", value))

# Identify invalid entries
invalid_months = clean[~clean['month'].apply(is_valid_month)]

# Display invalid entries
print("Invalid entries in 'month' column:")
print(invalid_months)

# Count invalid entries
print(f"Number of invalid entries: {len(invalid_months)}")

# Remove rows with invalid entries
clean[clean['month'].isin(valid_values)].copy()

# Verify the cleaning process
print("Unique values after cleaning:", clean['month'].unique())

Invalid entries in 'month' column:
     outcome     month  MMWR week     Age group Vaccine product  \
0       case  APR 2021     202114         12-17       all_types   
1       case  APR 2021     202114         18-29       all_types   
2       case  APR 2021     202114         30-49       all_types   
3       case  APR 2021     202114         50-64       all_types   
4       case  APR 2021     202114         65-79       all_types   
...      ...       ...        ...           ...             ...   
1586   death  AUG 2022     202235           80+       all_types   
1587   death  AUG 2022     202235  all_ages_adj       all_types   
1588   death  AUG 2022     202235  all_ages_adj         Janssen   
1589   death  AUG 2022     202235  all_ages_adj         Moderna   
1590   death  AUG 2022     202235  all_ages_adj          Pfizer   

      Vaccinated with outcome  Fully vaccinated population  \
0                           8                 3.688700e+04   
1                         674       

In [59]:
# checking for invalid age groups
unique_age_groups = clean['Age group'].unique()
print(unique_age_groups)

['12-17' '18-29' '30-49' '50-64' '65-79' '80+' 'all_ages_adj' '5-11']


In [60]:
# valid age group intervals 
expected_age_groups = unique_age_groups = [
    "12-17", "18-29", "30-49", "50-64", "65-79", "80+", "5-11", "all_ages_adj"
]

# Check if all values in the 'Age group' column are valid
unexpected_age_groups = clean[~clean['Age group'].isin(expected_age_groups)]

print("Unexpected Age group values:")
print(unexpected_age_groups['Age group'].unique())

# Number of unexpected age groups
print(f"Number of unexpected age groups: {len(unexpected_age_groups)}")

Unexpected Age group values:
[]
Number of unexpected age groups: 0


DUPLICATES

In [61]:
clean.duplicated()

0       False
1       False
2       False
3       False
4       False
        ...  
1586    False
1587    False
1588    False
1589    False
1590    False
Length: 1591, dtype: bool

In [62]:
clean.duplicated().sum()

np.int64(0)

In [63]:
duplicate_count = int(clean.duplicated().sum())
duplicate_count

0

DATETIME

In [64]:
clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1591 entries, 0 to 1590
Data columns (total 13 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   outcome                      1591 non-null   object 
 1   month                        1591 non-null   object 
 2   MMWR week                    1591 non-null   int64  
 3   Age group                    1591 non-null   object 
 4   Vaccine product              1591 non-null   object 
 5   Vaccinated with outcome      1591 non-null   int64  
 6   Fully vaccinated population  1591 non-null   float64
 7   Unvaccinated with outcome    1591 non-null   int64  
 8   Unvaccinated population      1591 non-null   float64
 9   Crude vax IR                 1591 non-null   float64
 10  Crude unvax IR               1591 non-null   float64
 11  Crude IRR                    1477 non-null   float64
 12  Continuity correction        1591 non-null   int64  
dtypes: float64(5), int

In [7]:
clean["month"] = pd.to_datetime(clean["month"], format='%m-%Y')

In [8]:
clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1591 entries, 0 to 1590
Data columns (total 13 columns):
 #   Column                       Non-Null Count  Dtype         
---  ------                       --------------  -----         
 0   outcome                      1591 non-null   object        
 1   month                        1591 non-null   datetime64[ns]
 2   MMWR week                    1591 non-null   int64         
 3   Age group                    1591 non-null   object        
 4   Vaccine product              1591 non-null   object        
 5   Vaccinated with outcome      1591 non-null   int64         
 6   Fully vaccinated population  1591 non-null   float64       
 7   Unvaccinated with outcome    1591 non-null   int64         
 8   Unvaccinated population      1591 non-null   float64       
 9   Crude vax IR                 1591 non-null   float64       
 10  Crude unvax IR               1591 non-null   float64       
 11  Crude IRR                    1477 non-null 

**CLEANING County_Level_of_community_Transmission.csv** (second dataset)

In [25]:
raw = pd.read_csv("Data/raw/County_Level_of_community_Transmission.csv")

In [26]:
print(raw)

           state_name        county_name  fips_code        date  \
0               Texas     Jackson County      48239  01/27/2020   
1            Nebraska     Buffalo County      31019  01/27/2020   
2            Illinois       Coles County      17029  01/27/2020   
3                Iowa         Ida County      19093  01/27/2020   
4           Tennessee        Knox County      47093  02/05/2020   
...               ...                ...        ...         ...   
3225217     Louisiana  Assumption Parish      22007  10/18/2022   
3225218  South Dakota        Hyde County      46069  10/18/2022   
3225219         Texas      Loving County      48301  10/18/2022   
3225220      Virginia     Lynchburg City      51680  10/18/2022   
3225221       Wyoming    Converse County      56009  10/18/2022   

        cases_per_100K_7_day_count_change  \
0                                   0.000   
1                                   0.000   
2                                   0.000   
3              

In [27]:
raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3225222 entries, 0 to 3225221
Data columns (total 7 columns):
 #   Column                                              Dtype  
---  ------                                              -----  
 0   state_name                                          object 
 1   county_name                                         object 
 2   fips_code                                           int64  
 3   date                                                object 
 4   cases_per_100K_7_day_count_change                   object 
 5   percent_test_results_reported_positive_last_7_days  float64
 6   community_transmission_level                        object 
dtypes: float64(1), int64(1), object(5)
memory usage: 172.2+ MB


In [73]:
raw.describe()

percent_test_results_reported_positive_last_7_days
count                                       2.823884e+06 
mean                                        1.154418e+01 
std                                         1.076036e+01 
min                                         0.000000e+00 
25%                                         3.790000e+00 
50%                                         8.640000e+00 
75%                                         1.618000e+01 
max                                         1.000000e+02

In [28]:
# raw.describe() does not calculate metrics in column cases_per_100K_7_day_count_change 
# because there are numerical and non-numerical entries in this column.
# so, convert non-numeric to numeric.

# Create a copy of the raw
clean = raw.copy()  
clean['cases_per_100K_7_day_count_change'] = pd.to_numeric(clean['cases_per_100K_7_day_count_change'], errors='coerce')

In [29]:
clean[['cases_per_100K_7_day_count_change', 'percent_test_results_reported_positive_last_7_days']].describe()

cases_per_100K_7_day_count_change  \
count                       2.468979e+06   
mean                        1.898672e+02   
std                         2.016008e+02   
min                        -9.018800e+02   
25%                         3.879500e+01   
50%                         1.277070e+02   
75%                         2.730585e+02   
max                         9.999970e+02   

       percent_test_results_reported_positive_last_7_days  
count                                       2.823884e+06   
mean                                        1.154418e+01   
std                                         1.076036e+01   
min                                         0.000000e+00   
25%                                         3.790000e+00   
50%                                         8.640000e+00   
75%                                         1.618000e+01   
max                                         1.000000e+02

Interpretation

Cases_per_100K_7_day_count_change

count: 2,468,979 valid non-null values. 
mean: The average 7-day case count change per 100,000 people is 189.87. 
std: The standard deviation is 201.60, indicating significant variability in the case count changes. 
min: The lowest change is -901.88 cases, which could indicate a large decrease in cases during a 7-day period. 
Percentiles: 25%: 38.80 (25% of the data is below this value). 50% (median): 127.71 (half the data is below this value). 
             75%: 273.06 (25% of the data is above this value). max: 999.997 cases per 100,000, likely indicating a very high increase.

Percent_test_results_reported_positive_last_7_days

count: 2,823,884 valid non-NaN values. 
mean: The average positivity rate over the last 7 days is 11.54%. 
std: The standard deviation is 10.76%, indicating some variation in positivity rates. 
min: The lowest positivity rate is 0%, indicating areas with no positive cases in the last 7 days. 
Percentiles: 25%: 3.79%. 50% (median): 8.64%. 75%: 16.18%. max: 100%, suggesting some areas are reporting exclusively positive results.

A positivity rate of 100% may indicate areas where testing is limited or only symptomatic individuals are being tested. 
The wide range from 0% to 100% suggests a significant disparity in how COVID-19 impacts different regions.

In [30]:
clean.to_csv("clean_data_2.csv", index=False)

In [3]:
clean = pd.read_csv("Data/clean/clean_data_2.csv")

In [ ]:
source = "clean_data_2.csv"
destination = "Data/clean/"
shutil.move(source, destination)

In [32]:
clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3225222 entries, 0 to 3225221
Data columns (total 7 columns):
 #   Column                                              Dtype  
---  ------                                              -----  
 0   state_name                                          object 
 1   county_name                                         object 
 2   fips_code                                           int64  
 3   date                                                object 
 4   cases_per_100K_7_day_count_change                   float64
 5   percent_test_results_reported_positive_last_7_days  float64
 6   community_transmission_level                        object 
dtypes: float64(2), int64(1), object(4)
memory usage: 172.2+ MB


MISSING AND INVALIDE VALUES

In [37]:
clean.isnull().sum()

state_name                                              2662
county_name                                             2662
fips_code                                                  0
date                                                       0
cases_per_100K_7_day_count_change                          0
percent_test_results_reported_positive_last_7_days    401338
community_transmission_level                           22661
dtype: int64

In [ ]:
# % of missing values in each column = number of missing values/total values * 100
clean.isnull().sum()/len(clean) * 100

In [38]:
# 'cases_per_100K_7_day_count_change' column has numerical, non-numerical and missing values.
# convert non-numerical to numerical and fillin missing values.

# standardize missing values and fillin
clean['cases_per_100K_7_day_count_change'] = clean['cases_per_100K_7_day_count_change'].replace(['', ' ', 'NA', 'N/A', 'null'], np.nan)
print(clean['cases_per_100K_7_day_count_change'].isnull().sum())

# make column totally numerical
mean_value = clean['cases_per_100K_7_day_count_change'].astype(float).mean()
clean['cases_per_100K_7_day_count_change'] = clean['cases_per_100K_7_day_count_change'].fillna(mean_value)
clean.to_csv('clean_data_2_updated.csv', index=False)


print(clean[clean['cases_per_100K_7_day_count_change'].isnull()]) 


0
Empty DataFrame
Columns: [state_name, county_name, fips_code, date, cases_per_100K_7_day_count_change, percent_test_results_reported_positive_last_7_days, community_transmission_level]
Index: []


In [43]:
# standardize missing values and fillin column 'percent_test_results_reported_positive_last_7_days'
clean['percent_test_results_reported_positive_last_7_days'] = clean[
'percent_test_results_reported_positive_last_7_days'].replace(['', ' ', 'NA', 'N/A', 'null'], np.nan)
print(clean['percent_test_results_reported_positive_last_7_days'].isnull().sum())

# make column totally numerical and fillin with median value.
median_value = clean['percent_test_results_reported_positive_last_7_days'].astype(float).mean()
clean['percent_test_results_reported_positive_last_7_days'] = clean['percent_test_results_reported_positive_last_7_days'].fillna(mean_value)
clean.to_csv('clean_data_2_updated.csv', index=False)


print(clean[clean['percent_test_results_reported_positive_last_7_days'].isnull()])


401338
Empty DataFrame
Columns: [state_name, county_name, fips_code, date, cases_per_100K_7_day_count_change, percent_test_results_reported_positive_last_7_days, community_transmission_level]
Index: []


'Data/clean/clean_data_2_updated.csv'

In [45]:
# Standardize missing values
clean['community_transmission_level'] = clean['community_transmission_level'].replace(['', ' ', 'NA', 'N/A', 'null'], np.nan)

# Check for missing values
print(clean['community_transmission_level'].isnull().sum())

# Fill missing values with the mode (most frequent value)
mode_value = clean['community_transmission_level'].mode().iloc[0]  # iloc in case there are more than one mode
clean['community_transmission_level'] = clean['community_transmission_level'].fillna(mode_value)

# Verify the column has no missing values
print(clean[clean['community_transmission_level'].isnull()])  # Should return an empty DataFrame

# Save the cleaned DataFrame to a CSV file
clean.to_csv('clean_data_2_updated.csv', index=False)

22661
Empty DataFrame
Columns: [state_name, county_name, fips_code, date, cases_per_100K_7_day_count_change, percent_test_results_reported_positive_last_7_days, community_transmission_level]
Index: []


In [47]:
# Standardize missing values
clean['state_name'] = clean['state_name'].replace(['', ' ', 'NA', 'N/A', 'null'], np.nan)

# Check for missing values
print(clean['state_name'].isnull().sum())

# Fill missing values with the mode (most frequent value)
mode_value = clean['state_name'].mode().iloc[0]  # iloc in case there are more than one mode
clean['state_name'] = clean['state_name'].fillna(mode_value)

# Verify the column has no missing values
print(clean[clean['state_name'].isnull()])  # Should return an empty DataFrame

# Save the cleaned DataFrame to a CSV file
clean.to_csv('clean_data_2_updated.csv', index=False)

2662
Empty DataFrame
Columns: [state_name, county_name, fips_code, date, cases_per_100K_7_day_count_change, percent_test_results_reported_positive_last_7_days, community_transmission_level]
Index: []


'Data/clean/clean_data_2_updated.csv'

In [50]:
# Standardize missing values
clean['county_name'] = clean['county_name'].replace(['', ' ', 'NA', 'N/A', 'null'], np.nan)

# Check for missing values
print(clean['county_name'].isnull().sum())

# Fill missing values with the mode (most frequent value)
mode_value = clean['county_name'].mode().iloc[0]  # iloc in case there are more than one mode
clean['county_name'] = clean['county_name'].fillna(mode_value)

# Verify the column has no missing values
print(clean[clean['county_name'].isnull()])  # Should return an empty DataFrame

# Save the cleaned DataFrame to a CSV file
clean.to_csv('clean_data_2_updated.csv', index=False)

2662
Empty DataFrame
Columns: [state_name, county_name, fips_code, date, cases_per_100K_7_day_count_change, percent_test_results_reported_positive_last_7_days, community_transmission_level]
Index: []


In [52]:
source = "clean_data_2_updated.csv"
destination = "Data/clean/"
shutil.move(source, destination)

'Data/clean/clean_data_2_updated.csv'

CONVERTING DATE

In [136]:
# convert date column into a datetime object.
clean['date'] = pd.to_datetime(clean['date'], format='%m/%d/%Y', errors='coerce')

In [139]:
# Because the % of missing values is low in state_name column, 
# We'll fillin with a placeholder 'unknown'instead of mode.
clean.fillna({"state_name": "unknown"}, inplace=True)

In [140]:
# Do the same for county_name column. 
clean.fillna({"county_name": "unknown"}, inplace=True)

In [5]:
# Filling column 'community_transmission_level' with a mode.
# we use iloc[0] just in case there are more than one mode.
mode_value = clean["community_transmission_level"].mode().iloc[0]
print(mode_value)
clean.fillna({"community_transmission_level": "mode_value"}, inplace=True)

high


CHECKING DUPLICATES

In [144]:
clean.duplicated().sum()
duplicat_count = int(clean.duplicated().sum())
duplicat_count

0

In [116]:
duplicated_rows = clean[clean.duplicated()]
print(duplicated_rows)

Empty DataFrame
Columns: [state_name, county_name, fips_code, date, cases_per_100K_7_day_count_change, percent_test_results_reported_positive_last_7_days, community_transmission_level]
Index: []
